README.txt
##
# JClic module for Moodle
#
# @package mod
# @subpackage jclic
# @copyright 2008 Departament d'Ensenyament de la Generalitat de Catalunya
# @source https://github.com/projectestac/moodle-mod_jclic/
# @license http://www.gnu.org/licenses/gpl-2.0.txt
##

Clic [1] is a set of free software for the development of multimedia educational 
activities.

This module allows the incorporation of JClic activities in Moodle. Its main 
features are:
- Allows embedding easily JClic activities in some Moodle course.
- Facilitates students tracing because it stores the score, date and duration 
of each of the attempts made ​​by the users.

-----------------------------------------------------------------------------
This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public
License as published by the Free Software Foundation; either
version 3 of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program (in a file called COPYING); if not, go
to http://www.gnu.org/copyleft/gpl.html or write to 

  Free Software Foundation, Inc.
  59 Temple Place - Suite 330
  Boston, MA 02111-1307 USA

-----------------------------------------------------------------------------
REQUIREMENTS
Compatible with Moodle 2.X releases

-----------------------------------------------------------------------------
INSTALL
Follow the usual installation instructions

-----------------------------------------------------------------------------
CREDITS

Developers:
   Sara Arjona Tellez (sarjona@xtec.cat)
   
With the support of:
    Education Government of Catalonia [2]
    Francesc Busquets Burguera (fbusquet@xtec.cat)



JClic community support:
   Monica Grau


Translations:
   Basque:
     Abel Camacho  - didaktika@santurtzieus.com
     Pedro Lonbide - zuzen@santurtzieus.com
     
   Greek:
     Yannis Kaskamanidis - kiolalis@gmail.com
     
   Galician:
     Xosé Luis Barreiro Cebey - xoseluis@edu.xunta.es
     
   Hebrew:
     Nadav Kavalerchik - nadavkav@gmail.com
 
   Brazilian
     Kleber Calegario Batista - klebercal@gmail.com


Special Thanks:
   Sandra Fernandez 
   Birgit Ferran
   Silvia Garcia Yague
   Albert Gasset Romo
   Eva Gea
   Toni Ginard
   David Monllao
   Jaume Rocarias



-----------------------------------------------------------------------------
[1] http://clic.xtec.cat/
[2] http://www.gencat.cat/ensenyament



CHANGES.txt
Release notes for JClic Module for Moodle (https://github.com/projectestac/moodle-mod_jclic/releases)

More information on each of the fixes can be found in the project
development home at https://github.com/projectestac/moodle-mod_jclic

Changes in progress
---------------------------------------------------------------------------------------
- Changed default report to detailed
- Redesigned edit form parameters
- Replaced .jclic.zip references in some translations

Changes in 3.0.1 - 2016061300 (2016/09/28)
---------------------------------------------------------------------------------------
- Added support for JClic HTML5 player

Changes in 3.0.0 - 2016052000 (2016/05/20)
---------------------------------------------------------------------------------------
- Deleted cron function
- Replaced deprecated add_intro_editor function
- Added first behat tests
- Prepared version for Moodle 3.1

Changes in 2.7.0 - 2015120300 (2015/12/03)
---------------------------------------------------------------------------------------
 - Changed version numbering method, now the first two numbers are the minimum Moodle required version
 - Solved two missing string identifiers
 - Solved string on reset
 - Changed links to lafarga

Changes in 0.2.6 - 2015061500 (2015/06/15)
---------------------------------------------------------------------------------------
 - Changed HTTP_RAW_POST_DATA by php://input
 - Requirements of Moodle 2.7 or greater
 - Revert Jarbase to http
 - Settings moved to config_plugins

Changes in 0.2.5 - 2015031300 (2015/03/13)
---------------------------------------------------------------------------------------
- Adapt module to Moodle 2.7 and 2.8
- Some navigation changes and added new events API classes
- Start testing new Travis CI scripts for continous integration
- Solved error selecting groups on results page
- Change status of no found session message
- Solved issue adding a jclic using a URL
- Updated translation to catalan and spanish
- Set JClic default settings to https
- Add protocol recognision through wwwroot url for the applet
- Removed support for versions before 2.6

Changes in 0.2.4 - 2014052901 (2014/11/07)
---------------------------------------------------------------------------------------
- Adapted module to Moodle 2.6
- Solved error with activities that lasts more than 99 seconds saving total_time in seconds instead of miliseconds. Now the limit is about 27 hours per activity
- Changed default jclicplugin.js (to use the clic server one) and updated the local to the last version
- Reviewed permissions for lang/pt_br/jclic.php file


Changes in 0.2.3 - 2013022200 (2013/09/19)
---------------------------------------------------------------------------------------
- Fixed bug when deleting jclic activities and enableajax is enabled
- Fixed bug when upgrading empty maxgrade (reported by Ferran Recio - UPCnet)
- Fixed bug when editing title from main course page (is still not working but solve some minor warnings)
- Jclicplugin.js can be loaded from config.


Changes in 0.2.3 RC - 2013022200 (2013/02/22)
---------------------------------------------------------------------------------------
- Added Brazilian translation
- Updated Euskara translation (downloaded from AMOS)
- Reviewed English translation (thanks to German Valero!)
- Reviewed some SQL to avoid Postgre problem with userid
- Included SVG icons (thanks to Pau Ferrer!)
- Fixed bug with users reporting (only first page was showed)


Changes in 0.2.2 (2012/08/24)
---------------------------------------------------------------------------------------
- Added Brazilian translation
- Added Galician translation
- Fixed some bugs related with importation from Moodle 1.9
- Solved some 'Strict standards' warnings
- Adapted code for Moodle 2.3 (like addinstance capability)


Changes in 0.2.1 (2012/06/05)
---------------------------------------------------------------------------------------
- First version of the jclic module for Moodle 2.


Changes in 0.1.0.11 (2011/05/25)
---------------------------------------------------------------------------------------
- Added language parameter to the activity edition page to let specify the player language (http://projectes.lafarga.cat/projects/jclicmoodle/tracker/291/1641/detail)
- Added exiturl parameter to the activity edition page to let specify exit URL (http://projectes.lafarga.cat/projects/jclicmoodle/tracker/291/1287/detail)
- Revised information showed in summary
- Solved problem when moodle is installed in the root (without path in the URL)
- Added jclic_scale_used_anywhere function sended by David Monllao from URV to improve the performance of the module
- Fixed warnings when restoring jclic activities (http://projectes.lafarga.cat/projects/jclicmoodle/tracker/288/1200/detail)
- Fixed bug with single quote reported to Moodle tracker (http://tracker.moodle.org/browse/CONTRIB-2079?page=com.atlassian.jira.plugin.system.issuetabpanels%3Acvs-tabpanel)
- Fixed warning message when there were no users in the course (http://projectes.lafarga.cat/projects/jclicmoodle/tracker/288/1335/detail)

lib.php
<?PHP  // $Id: lib.php,v 1.19 2011-05-25 12:13:03 sarjona Exp $

/// Library of functions and constants for module jclic

require_once($CFG->libdir.'/pagelib.php');


if (!isset($CFG->jclic_jclicpluginjs)) {
    set_config("jclic_jclicpluginjs", "http://clic.xtec.cat/dist/jclic/jclicplugin.js");
}
if (!isset($CFG->jclic_lap)) {
    set_config("jclic_lap", "5");
}

function jclic_add_instance($jclic) {
/// Given an object containing all the necessary data, 
/// (defined by the form in mod.html) this function 
/// will create a new instance and return the id number 
/// of the new instance.
    $jclic->url=trim($jclic->url);
    if ($jclic->skin=='') $jclic->skin="default";

    return insert_record("jclic", $jclic);
}


function jclic_update_instance($jclic) {
/// Given an object containing all the necessary data, 
/// (defined by the form in mod.html) this function 
/// will update an existing instance with new data.
    $jclic->id = $jclic->instance;
    $jclic->url=trim($jclic->url);
    if(update_record("jclic", $jclic)){
   		return jclic_update_gradebook(null, $jclic);
    }
	return false;
}


function jclic_delete_instance($id) {
/// Given an ID of an instance of this module, 
/// this function will permanently delete the instance 
/// and any data that depends on it.


    if (! $jclic = get_record("jclic", "id", "$id")) {
        return false;
    }
    $result = true;
    
    # Delete any dependent records here #
    $rs =  get_record("jclic_sessions", "jclicid", "$id");
    foreach($rs as $session){
      delete_records("jclic_activities", "session_id", "$rs->session_id");
	  }

    delete_records("jclic_sessions", "jclicid", "$id");
    
    if (! delete_records("jclic", "id", "$id")) {
        $result = false;
    }
    
    
    return $result;
}

function jclic_user_outline($course, $user, $mod, $jclic) {
/// Return a small object with summary information about what a 
/// user has done with a given particular instance of this module
/// Used for user activity reports.

	$jclic_summary = jclic_get_sessions_summary($jclic->id, $user->id);
	$return->time = $jclic_summary->totaltime;
	$maxgrade = 0;
	if (property_exists($jclic, 'maxgrade')) $maxgrade = $jclic->maxgrade;
	$return->info = get_string("score", "jclic").": ".$jclic_summary->score."/".$maxgrade;
    return $return;
}

function jclic_user_complete($course, $user, $mod, $jclic) {
/// Print a detailed representation of what a  user has done with 
/// a given particular instance of this module, for user activity reports.

    return true;
}

function jclic_print_recent_activity($course, $isteacher, $timestart) {
/// Given a course and a time, this module should find recent activity 
/// that has occurred in jclic activities and print it out.
/// Return true if there was output, or false is there was none.

    global $CFG;

    return false;  //  True if anything was printed, otherwise false 
}

function jclic_cron () {
/// Function to be run periodically according to the moodle cron
/// This function searches for things that need to be done, such 
/// as sending out mail, toggling flags etc ... 

    global $CFG;

    return true;
}

function jclic_grades($jclicid, $userid=NULL) {
	global $CFG;
/// Must return an array of grades for a given instance of this module, 
/// indexed by user.  It also returns a maximum allowed grade.
///
///    $return->grades = array of grades;
///    $return->maxgrade = maximum allowed grade;
///

    if (!$jclic = get_record('jclic', 'id', $jclicid)) {
        return NULL;
    }

    $return = new object();
	$grades = array();
	if ($userid==NULL){
  	 if ($participants = jclic_get_participants($jclicid)){
  		foreach ($participants as $participant) {
				$grades = jclic_user_grades($grades, $jclic, $participant->userid);
  		}
   	 }
	} else{
		$grades = jclic_user_grades($grades, $jclic, $userid);
	}
  	$return->grades=$grades;
  	$return->maxgrade=$jclic->maxgrade;
   return $return;


}

function jclic_user_grades($grades, $jclic, $userid) {
	$sessions_summary=jclic_get_sessions_summary($jclic->id, $userid);
	if ($jclic->avaluation=='score'){
		$grades[$userid]=$sessions_summary->score;				
	}else{
		$grades[$userid]=$sessions_summary->solved;
	}
	return $grades;
}

function jclic_update_gradebook($jclic_activity=NULL, $jclic=NULL){
	global $CFG;

    if (!function_exists('grade_update')) { //workaround for buggy PHP versions
	    if (file_exists($CFG->libdir.'/gradelib.php')) include_once($CFG->libdir.'/gradelib.php');
    }
	
    if (!function_exists('grade_update')) {
	    return true;
    }

	if ($jclic_activity != NULL){
	    if (!$jclic_session = get_record('jclic_sessions', 'session_id', $jclic_activity->session_id)) {
	        return NULL;
	    }
	    if (!$jclic = get_record('jclic', 'id', $jclic_session->jclicid)) {
	        return NULL;
	    }
	}
	if ($jclic != NULL){
		// Get the grades
		$userid = (isset($jclic_session)?$jclic_session->user_id:null);
		$jclic_grades = jclic_grades($jclic->id, $userid);
		$grades = array();
		foreach ($jclic_grades->grades as $k=>$v){
			$grades[$k]->userid=$k;
			$grades[$k]->rawgrade=$v;			
		}
		
		// Get the params
	    $params = array('itemname'=>$jclic->name);
	    $params['gradetype'] = GRADE_TYPE_VALUE;
	    if ($jclic->maxgrade==NULL) $jclic_grades->maxgrade = 0;
	    $params['gradepass']  = $jclic_grades->maxgrade;
	    $params['grademin']  = 0;

		if ($jclic->avaluation=='score') $params['grademax']  = 100;
		else $params['grademax']  = $jclic_grades->maxgrade;

		grade_update('mod/jclic', $jclic->course, 'mod', 'jclic', $jclic->id, 0, $grades, $params);	
		return true;
	}
	return false;
}

function jclic_get_students($cm, $course, $jclicid) {
	global $CFG;
	$version_moodle = (float)substr($CFG->release,0,3);
	if ($version_moodle>=1.7){
    $context = get_context_instance(CONTEXT_MODULE, $cm->id);
    $currentgroup = get_and_set_current_group($course, groupmode($course, $cm));
    $users = get_users_by_capability($context, 'mod/jclic:submit', 'u.id, u.id', '', '', '', $currentgroup, '', false);
    $dbstudents = array();
    if (!empty($users)){
        $select = 'SELECT u.id AS userid, u.firstname, u.lastname, u.picture ';
        $sql = 'FROM '.$CFG->prefix.'user u '.
               'LEFT JOIN '.$CFG->prefix.'jclic_sessions a ON u.id = a.user_id AND a.jclicid = '.$jclicid.' '.
               'WHERE u.id IN ('.implode(',', array_keys($users)).') ';
        $dbstudents=get_records_sql($select.$sql);
    }
  }else{
  	$dbstudents = get_records_sql("SELECT DISTINCT us.userid, u.firstname, u.lastname
  				       FROM {$CFG->prefix}user u,{$CFG->prefix}user_students us, {$CFG->prefix}jclic j
  				       WHERE us.course=j.course AND j.id=$jclicid AND u.id=us.userid");
  }
  return $dbstudents;
}

function jclic_get_participants($jclicid) {
//Must return an array of user records (all data) who are participants
//for a given instance of jclic. Must include every user involved
//in the instance, independient of his role (student, teacher, admin...)
//See other modules as example.

    global $CFG;

    //Get students
    $students = get_records_sql("SELECT DISTINCT u.id, u.id as userid
                                 FROM {$CFG->prefix}user u,
                                      {$CFG->prefix}jclic_sessions js
                                 WHERE js.jclicid = '$jclicid' and
                                       u.id = js.user_id");
    //Get teachers
    $teachers = get_records_sql("SELECT DISTINCT u.id, u.id as userid
                                 FROM {$CFG->prefix}user u,
                                      {$CFG->prefix}jclic_sessions js
                                 WHERE js.jclicid = '$jclicid' and
                                       u.id = js.user_id");

    //Add teachers to students
    if ($teachers) {
        foreach ($teachers as $teacher) {
            $students[$teacher->id] = $teacher;
        }
    }
    //Return students array (it contains an array of unique users)
    return ($students);
}

function jclic_scale_used ($jclicid,$scaleid) {
//This function returns if a scale is being used by one jclic
//it it has support for grading and scales. Commented code should be
//modified if necessary. See forum, glossary or journal modules
//as reference.
   
    $return = false;

    return $return;
}

/**
 * Checks if scale is being used by any instance of jclic
 *
 * This is used to find out if scale used anywhere
 * @param $scaleid int
 * @return boolean
 */
function jclic_scale_used_anywhere($scaleid) {
   return false;
}



//////////////////////////////////////////////////////////////////////////////////////

/**
* Get user sessions
*
* @return array			[0=>session1,1=>session2...] where session1 is an array with keys: id,score,totaltime,starttime,done,solved,attempts. First sessions are newest.
* @param object $jclicid	The jclic to get sessions
* @param object $userid		The user id to get sessions
*/
function jclic_get_sessions($jclicid, $userid) {
	global $CFG;

  $sessions=array();
	if($rs = get_records_sql("SELECT js.*
                            FROM {$CFG->prefix}jclic j, {$CFG->prefix}jclic_sessions js 
                            WHERE j.id=js.jclicid AND js.jclicid=$jclicid AND js.user_id=$userid
                            ORDER BY js.session_datetime")){
		$i=0;
		foreach($rs as $session){
			$activity=jclic_get_activity($session);
			$activity->attempts=$i+1;
			$sessions[$i++]=$activity;
		}
	}
	return $sessions;
}

/**
* Get session activities
*
* @return array			[0=>act0,1=>act1...] where act0 is an array with keys: activity_id,activity_name,num_actions,score,activity_solved,qualification, total_time. First activity are oldest.
* @param string $session_id		The session id to get actitivies
*/
function jclic_get_activities($session_id) {
	global $CFG;
	
	if($rs = get_records("jclic_activities", "session_id", "$session_id",'activity_id')){
		$i=0;
		foreach($rs as $activity){
			$activities[$i++]=$activity;
		}
	}
	return $activities;
}

/**
* Get information about activities of specified session
*
* @return array			Array has these keys id,score,totaltime,starttime,done,solved,attempts
* @param object $session	The session object
*/
function jclic_get_activity($session) {
	global $CFG;
	$activity->starttime=$session->session_datetime;
	$activity->session_id=$session->session_id;
	if($rs = get_record_sql("SELECT AVG(ja.qualification) as qualification, SUM(ja.total_time) as totaltime
    				 FROM {$CFG->prefix}jclic_activities ja 
				 WHERE ja.session_id='$session->session_id'")){
		$activity->score=round($rs->qualification,0);
		$activity->totaltime=round($rs->totaltime/60000,0)."' ".round(fmod($rs->totaltime,60000)/1000,0)."''";
	}
	if ($rs = get_record_sql("SELECT COUNT(*) as done
                            FROM (SELECT DISTINCT ja.activity_name 
                                  FROM  {$CFG->prefix}jclic_activities ja 
                                  WHERE ja.session_id='$session->session_id') t")){
    $activity->done=$rs->done;
  }
	if ($rs = get_record_sql("SELECT COUNT(*) as solved
                            FROM (SELECT DISTINCT ja.activity_name 
                                  FROM {$CFG->prefix}jclic_activities ja 
                                  WHERE ja.session_id='$session->session_id' AND ja.activity_solved=1) t")){
    $activity->solved=$rs->solved;
	}

	return $activity;
}

/**
* Get user activity summary
*
* @return object	session object with score, totaltime, activities done and solved and attempts information
*/
function jclic_get_sessions_summary($jclicid, $userid) {
	global $CFG;
	
	$sessions_summary = new stdClass;
  	if ($rs = get_record_sql("SELECT COUNT(*) AS attempts, AVG(t.qualification) AS qualification, SUM(t.totaltime) AS totaltime, MAX(t.starttime) AS starttime
                            FROM (SELECT AVG(ja.qualification) AS qualification, SUM(ja.total_time) AS totaltime, MAX(js.session_datetime) AS starttime
                                  FROM {$CFG->prefix}jclic j, {$CFG->prefix}jclic_sessions js, {$CFG->prefix}jclic_activities ja  
                                  WHERE j.id=js.jclicid AND js.user_id=$userid AND js.jclicid=$jclicid AND ja.session_id=js.session_id
                                  GROUP BY js.session_id) t")){
		$sessions_summary->attempts=$rs->attempts;
		$sessions_summary->score=round($rs->qualification,0);
		$sessions_summary->totaltime=round($rs->totaltime/60000,0)."' ".round(fmod($rs->totaltime,60000)/1000,0)."''";
		$sessions_summary->starttime=$rs->starttime;
	}
	
	if ($rs = get_record_sql("SELECT COUNT(*) as done
                            FROM (SELECT DISTINCT ja.activity_name 
                                  FROM {$CFG->prefix}jclic j, {$CFG->prefix}jclic_sessions js, {$CFG->prefix}jclic_activities ja 
                                  WHERE j.id=js.jclicid AND js.user_id=$userid AND js.jclicid=$jclicid AND js.session_id=ja.session_id)  t")){
		$sessions_summary->done=$rs->done;
	}
	if ($rs = get_record_sql("SELECT COUNT(*) as solved
                            FROM (SELECT DISTINCT ja.activity_name 
                                  FROM {$CFG->prefix}jclic j, {$CFG->prefix}jclic_sessions js, {$CFG->prefix}jclic_activities ja 
                                  WHERE j.id=js.jclicid AND js.user_id=$userid AND js.jclicid=$jclicid AND js.session_id=ja.session_id AND ja.activity_solved=1) t")){
    $sessions_summary->solved=$rs->solved;
	}
	return $sessions_summary;
}

/**
* Format time from milliseconds to string 
*
* @return string Formated string [x' y''], where x are the minutes and y are the seconds.	
* @param int $time	The time (in ms)
*/
function jclic_time2str($time){
  return round($time/60000,0)."' ".round(fmod($time,60000)/1000,0)."''";
}

/**
* Print data in array as a row. Uses jclic_make_row function to get the string representation to print
*
* @param array $row Data
* @param array $align Cell alignment
* @param array $props Style, rowspan, colspan...
* @param boolean $header If true the row will be printed as th; otherwise as td.
*/
function jclic_print_row($row, $align, $props=array(), $header=false){
  echo jclic_make_row($row, $align, $props, $header);
}

function jclic_make_row($row, $row_align, $row_props=array(), $header=false){
  $strrow = '';
  $strtd = $header?'th':'td';
  if (sizeof($row)>0){
    $strrow.='<tr>';
    $i=0;
		foreach($row as $cell){
		  $prop = ''; $align='';
		  if ($row_props!='' && array_key_exists($i, $row_props) && $row_props[$i]!='') $prop = ' '.$row_props[$i].' ';
		  if ($row_align!=''&& array_key_exists($i, $row_align) && $row_align[$i]!='') $align = ' align='.$row_align[$i].' ';
      $strrow.='<'.$strtd.' '.($header?'':'class="cell"').$prop.$align.'>'.$cell.'</'.$strtd.'>';
      $i++;	  
    }
    $strrow.='</tr>';
  }  
  return $strrow;
}

/**
* Print a table data with all session activities 
* 
* @param string $session_id The session identifier
*/
function print_session_activities($session_id){
  // Import language strings
  $stractivity = get_string("activity", "jclic");
  $strsolved = get_string("solved", "jclic");
  $stractions = get_string("actions", "jclic");
  $strtime = get_string("time", "jclic");
  $strscore  = get_string("score", "jclic");
  $stryes = get_string("yes");
  $strno = get_string("no");
  
  // Print activities for each session
  $activities=jclic_get_activities($session_id);    
	if (sizeof($activities)>0){
	  echo '<tr><td colspan="7"><div id="'.$session_id.'" style="display:none;visible:hidden">';
	  echo '<table class="generaltable" align="center" border="0" cellpadding="5" cellspacing="1" width="80%"><tbody>';
	  $subtable_align=array('left','center','center','center','center');
	  jclic_print_row(array($stractivity, $strsolved, $stractions, $strtime, $strscore),$subtable_align,'',true);
		foreach($activities as $activity){
		  $subtable_props=array('', 'style="background-color:'.($activity->activity_solved?'#DFFFDF':'#FFDFDF').'"','','','');
		  $act_percent=$activity->num_actions>0?round(($activity->score/$activity->num_actions)*100,0):0;
		  jclic_print_row(array($activity->activity_name, ($activity->activity_solved?$stryes:$strno), $activity->score.'/'.$activity->num_actions.' ('.$act_percent.'%)', jclic_time2str($activity->total_time), $activity->qualification.'%'), $subtable_align, $subtable_props);
		}
		echo '</table></div></td></tr>';
  }
}

function jclic_get_skins(){
  return array('@default.xml' => 'default','@blue.xml' => 'blue','@orange.xml' => 'orange','@green.xml' => 'green','@simple.xml' => 'simple', '@mini.xml' => 'mini');
} 


/**
* Get an array with the languages
*
* @return array   The array with each language.
*/
function jclic_get_languages(){
    global $CFG;
    $tmplanglist = get_list_of_languages();
    $langlist = array();
    foreach ($tmplanglist as $lang=>$langname) {
        if (substr($lang, -5) == '_utf8') {   //Remove the _utf8 suffix from the lang to show
            $lang = substr($lang, 0, -5);
        }
        $langlist[$lang]=$langname;
    }
    return $langlist;
}


/**
* Get moodle server
*
* @return string                myserver.com:port
*/
function jclic_get_server() {
    global $CFG;

    if (!empty($CFG->wwwroot)) {
        $url = parse_url($CFG->wwwroot);
    }

    if (!empty($url['host'])) {
        $hostname = $url['host'];
    } else if (!empty($_SERVER['SERVER_NAME'])) {
        $hostname = $_SERVER['SERVER_NAME'];
    } else if (!empty($_ENV['SERVER_NAME'])) {
        $hostname = $_ENV['SERVER_NAME'];
    } else if (!empty($_SERVER['HTTP_HOST'])) {
        $hostname = $_SERVER['HTTP_HOST'];
    } else if (!empty($_ENV['HTTP_HOST'])) {
        $hostname = $_ENV['HTTP_HOST'];
    } else {
        notify('Warning: could not find the name of this server!');
        return false;
    }

    if (!empty($url['port'])) {
        $hostname .= ':'.$url['port'];
    } else if (!empty($_SERVER['SERVER_PORT'])) {
        if ($_SERVER['SERVER_PORT'] != 80 && $_SERVER['SERVER_PORT'] != 443) {
            $hostname .= ':'.$_SERVER['SERVER_PORT'];
        }
    }

    return $hostname;
}

/**
* Get moodle path
*
* @return string                /path_to_moodle
*/
function jclic_get_path() {
    global $CFG;

	$path = '/';
    if (!empty($CFG->wwwroot)) {
        $url = parse_url($CFG->wwwroot);
		if (array_key_exists('path', $url)){
			$path = $url['path'];			
		}
    }
    return $path;
}

if (!function_exists('get_coursemodule_from_id')) {
	// Add this function for Moodle < 1.6
	function get_coursemodule_from_id($modulename, $cmid, $courseid=0) {
		global $CFG;
		return get_record_sql("
			SELECT 
				cm.*, m.name, md.name as modname
			FROM 
				{$CFG->prefix}course_modules cm,
				{$CFG->prefix}modules md,
				{$CFG->prefix}$modulename m
			WHERE 
				".($courseid ? "cm.course = '$courseid' AND " : '')."
				cm.id = '$cmid' AND
				cm.instance = m.id AND
				md.name = '$modulename' AND
				md.id = cm.module
		");
	}
}

/**
 * This function is used by the reset_course_userdata function in moodlelib.
 * This function will remove all jclic sessions in the specified course.
 * @param $data the data submitted from the reset course.
 * @return array status array
 */
function jclic_reset_userdata($data) {
    global $CFG;
    require_once($CFG->libdir.'/filelib.php');

    $status = array();
    if (!empty($data->reset_jclic_deleteallsessions)) {
        $select = 'session_id IN'
            . " (SELECT s.session_id FROM {$CFG->prefix}jclic_sessions s"
            . " INNER JOIN {$CFG->prefix}jclic j ON s.jclicid = j.id"
            . " WHERE j.course = {$data->courseid})";
        delete_records_select('jclic_activities', $select);

        $select = 'jclicid IN'
            . " (SELECT j.id FROM {$CFG->prefix}jclic j"
            . " WHERE j.course = {$data->courseid})";
        delete_records_select('jclic_sessions', $select);

        $status[] = array('component' => get_string('modulenameplural', 'jclic'),
                          'item' => get_string('deleteallsessions', 'jclic'),
                          'error' => false);
    }
    return $status;
}

/**
 * Called by course/reset.php
 * @param $mform form passed by reference
 */
function jclic_reset_course_form_definition(&$mform) {
    $mform->addElement('header', 'jclicheader', get_string('modulenameplural', 'jclic'));

    $mform->addElement('checkbox', 'reset_jclic_deleteallsessions', get_string('deleteallsessions', 'jclic'));
}

/**
 * Course reset form defaults.
 */
function jclic_reset_course_form_defaults($course) {
    return array('reset_jclic_deleteallsessions' => 1);
}

?>
